In [75]:
import pandas as pd
import numpy as np

## 1.  Import the package to help with following analysis process

In [76]:
# upload the step count data and store it into dataframe
df = pd.read_csv('person1-step-data.csv',index_col = 1,parse_dates=True)

## 2. Upload the data and change the datatype in the cvs file to the datatype of datetime.

When transfer the Date's datatype to datetime, it really help us to group the data by years or months and bring a huge improvement in efficient aspect. 

In [77]:
# observe the data in the df
df

Source  Hour  Count
Date                                      
2014-12-07  Person1 iPhone SE     8     13
2014-12-07  Person1 iPhone SE     8     13
2014-12-07  Person1 iPhone SE     8      1
2014-12-07  Person1 iPhone SE     8      9
2014-12-07  Person1 iPhone SE     8     15
...                       ...   ...    ...
2021-09-22  Person1 iPhone SE    17   1241
2021-09-22  Person1 iPhone SE    17   1212
2021-09-22  Person1 iPhone SE    18    808
2021-09-22  Person1 iPhone SE    18    392
2021-09-22  Person1 iPhone SE    18     34

[183787 rows x 3 columns]

## 3. To check and observe the nature of data in the dataframe we create

We try to do the operation about examin whether the null values exsits in the dataframe

In [78]:
# to check whether exist NULL/missing value in the dataframe
df.isnull().any().any()

False

In all the cells in the dataframe,there do not have the NULL value

In [79]:
# observe the data in count column
df["Count"].describe()

count    183787.000000
mean         79.791558
std         358.099285
min           1.000000
25%           6.000000
50%          17.000000
75%          35.000000
max       88359.000000
Name: Count, dtype: float64

To get the math feature and range of these data in the dataframe, and check whether there are any outliers should be cleaned. From above, it result is pretty normal,though the max step count is reach 88359 in one certain day, however the data is puasible.

## 4. we integrate the data and resample the data by months

In [80]:
df = df.drop(columns=['Hour'])
df = df.resample('M').sum()
df

Count
Date              
2014-12-31  111590
2015-01-31  131321
2015-02-28  106851
2015-03-31  101621
2015-04-30   91726
...            ...
2021-05-31  518898
2021-06-30  406440
2021-07-31  360565
2021-08-31  321087
2021-09-30  205743

[82 rows x 1 columns]

The driving question is try to find whether the tester is more active in the last 12 months than the 12 months before that. However, we find for 2014 and 2021, the data is imcompleted, there are only one month and nine months data relatively in 2214 and 2021. Then we try to drop these data belong to 2014 and 2021, and analysis the completed data from 2015 to 2020 **(6 years)**

In [81]:
df_6_years = df['2015':'2020']
df_6_years

Count
Date              
2015-01-31  131321
2015-02-28  106851
2015-03-31  101621
2015-04-30   91726
2015-05-31  177845
...            ...
2020-08-31   92018
2020-09-30   86170
2020-10-31   88564
2020-11-30   94503
2020-12-31  117473

[72 rows x 1 columns]

Only extract the data from 2015 to 2020

**Calculalte the total step count for each month by resampling skill**

In [82]:
df_6_years = df_6_years.resample('M').sum()
df_6_years


Count
Date              
2015-01-31  131321
2015-02-28  106851
2015-03-31  101621
2015-04-30   91726
2015-05-31  177845
...            ...
2020-08-31   92018
2020-09-30   86170
2020-10-31   88564
2020-11-30   94503
2020-12-31  117473

[72 rows x 1 columns]

## 5. Data visualization

create a new dataframe and make the years as the columns in this dataframe

In [83]:
df1 = pd.DataFrame(data=df_6_years.index, columns=['Date'])
df2 = pd.DataFrame(data=df_6_years.values, columns=['Count'])
df_new = pd.merge(df1, df2, left_index=True, right_index=True)
year1 = df_new.iloc[0:12,1].values
year2 = df_new.iloc[12:24,1].values
year3 = df_new.iloc[24:36,1].values
year4 = df_new.iloc[36:48,1].values
year5 = df_new.iloc[48:60,1].values
year6 = df_new.iloc[60:72,1].values
date = ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sept','Oct','Nov','Dec',]

Store the date in the series and use the dic to contain these series, then transfer the data into the dateframe which called **df_final**

In [84]:
data = {
    'Month':pd.Series(date),
    '2015':pd.Series(year1),
    '2016':pd.Series(year2),
    '2017':pd.Series(year3),
    '2018':pd.Series(year4),
    '2019':pd.Series(year5),
    '2020':pd.Series(year6),
     
}
df_final = pd.DataFrame(data)

We try to use the plotly package in python to acheive interact plot,user can click the year options on the right side of the plot, they can decide the certain year should be demostrated on the plot 

In [85]:
import plotly.express as px
fig = px.line()  
for i,n in enumerate(df_final.columns[1:7]):  
  fig.add_scatter(x=df_final['Month'], y= df_final[df_final.columns[i+1]], name= df_final.columns[i+1]) 

fig.update_layout(  
title = 'Monthly step count'
,xaxis_title = 'Months'  
,yaxis_title = 'total step count'  
,font = dict(size = 25)  
,template = 'plotly_dark' #"plotly", "plotly_white", "plotly_dark", "ggplot2", "seaborn", "simple_white", "none"  
)  
fig.show()